Add RDataFrame to the primer (#724)
root-project/web: primer/index.md

    TFile in_file("my_rootfile.root");

    // Get the Histogram out
    auto h = in_file.Get<TH1F>("my_histogram");

    // Draw it
    auto myCanvas = new TCanvas();
    h->DrawClone();
}
{% endhighlight %}

## The ROOT dataset

> ROOT offers its own data format with the TTree class. These are some of its advantages:

Reviewer: Given that the ROOT file format is independent from TTree, I would suggest to rephrase this as:
```suggestion
ROOT can handle storage/retrieval of large datasets through the {% include ref class="TTree" %} class. These are some of its advantages:
```

- Store any type of C++ object in the columns.

ROOT automatically applies compression algorithms on TTree to reduce
the memory consumption. A value that is in most cases the same will
consume only small space on your disk (but it has to be decompressed on
reading). Nevertheless, you should think about the design of your
dataset and your analyses as soon as the processing time exceeds some
minutes.

- Try to keep a TTree simple and use appropriate variable types.
  If your measurement has only a limited precision, it is needless to
  store it with double precision.
- Experimental conditions that do not change with every single
  measurement should be stored in a separate tree. Although the
  compression can handle redundant values, the processing time
  increases with every variable that has to be filled.
- The function SetCacheSize(long) specifies the size of the cache
  for reading a TTree object from a file. The default value is 30MB.
  A manual increase may help in certain situations. Please note that
  the caching mechanism can cover only one TTree object per TFile
  object.
- If you would like to measure I/O performance when reading a dataset you could

>   use the TTreePerfStats class. The ROOT documentation on this class also

Reviewer: ```suggestion
  use the {% include ref class="TTreePerfStats" %} class. The ROOT documentation on this class also
```

other useful statistics. As an example, imagine your task now consists in
finding the relations among the variables, without knowing the code used to
generate them. You can read back the stored dataset with RDataFrame as follows:

{% highlight C++ %}
root[0] ROOT::RDataFrame df{"cond_data","conductivity_experiment.root"};
{% endhighlight %}

Next, try to type the following at the prompt:

{% highlight C++ %}
root[1] auto g1 = df.Graph("current","potential");
root[2] g1->Draw("ap");
{% endhighlight %}

You just produced a correlation plot with RDataFrame! From this point, try to
produce a plot after having filtered some events out of the dataset:

{% highlight C++ %}
root [3] auto g2 = df.Filter("temperature<270").Graph("current","potential");
root [4] g2->Draw("ap");
{% endhighlight %}

Or maybe you can try producing a plot out of a newly defined variable starting
from the available columns:

{% highlight C++ %}
root [5] auto g3 = df.Define("curoverpot", "current/potential").Graph("curoverpot","temperature");
root [6] g3->Draw("ap");
{% endhighlight %}

## Working with columns that hold collections

RDataFrame reads collections as the special type

> [ROOT::RVec](https://root.cern/doc/master/classROOT_1_1VecOps_1_1RVec.html):

Reviewer: ```suggestion
[ROOT::VecOps::RVec](https://root.cern/doc/master/classROOT_1_1VecOps_1_1RVec.html):
```

Reviewer 2: I'd rather advertise it as `ROOT::RVec` here (it's valid)

for example, a column where each element is an array of floating point numbers

> can be read as a ROOT::RVecF. C-style arrays (with variable or static size),

Reviewer: Given that we talked about the `RVec` type (and not `RVecF`), I would prefer to stick to that, i.e.
```suggestion
can be read as a `ROOT::VecOps::RVec<float>`. C-style arrays (with variable or static size),
```

Reviewer 2: we should advertise the short aliases though

Author: Maybe I can write a short sentence describing the existence of the aliases

Author: I've added a small sentence there. If it looks ok this can be merged

Reviewer 2: I'd like to figure out what to do with the other versions of the primer before merging

STL vectors and most other collection types can be read this way.

> RVec is a container similar to std::vector (and can be used just like a
> std::vector) but it also offers a rich interface to operate on the array

Reviewer: ```suggestion
`RVec` is a container similar to `std::vector` (and can be used just like a
`std::vector`) but it also offers a rich interface to operate on the array
```

elements in a vectorised fashion, similarly to Python's NumPy arrays.

For example, to fill a histogram with the "pt" of selected particles for each

> event, Define() can be used to create a column that contains the desired array

Reviewer: ```suggestion
event, `Define()` can be used to create a column that contains the desired array
```

elements as follows:

{% highlight C++ %}
// h is filled with all the elements of `good_pts`, for each event
auto h = df.Define("good_pts", [](const ROOT::RVecF &pt) { return pt[pt > 0]; })
           .Histo1D("good_pts");
{% endhighlight %}

## Parallel analysis with RDataFrame

RDataFrame offers the possibility to exploit multicore machines by parallelizing
the computations thanks to implicit multi-threading. Each thread will get a
different part of the entries in the original dataset, divided fairly among all
threads. The only modification required to your code is the addition of the
following line *before* constructing the RDataFrame object:

{% highlight C++ %}
ROOT::EnableImplicitMT();
{% endhighlight %}

Simple as that. More details are given
[in the RDataFrame documentation](https://root.cern/doc/master/classROOT_1_1RDataFrame.html#parallel-execution).

## Using RDataFrame in a Python application

The RDataFrame class can be just as well used in a Python script, thanks to the
dynamic C++/Python translation offered by PyROOT. The interface is in general
the same as when writing C++ applications:

> {% highlight Python %}

Reviewer: ```suggestion
{% highlight Python %}
import ROOT
```
